source/fe/fe_enriched.cc

```diff
@@ -0,0 +1,927 @@
+{
+  UpdateFlags out = fe_system.requires_update_flags(flags);
+
+  if (is_enriched)
+    {
+      // if we ask for values or gradients, then we would need quadrature points
+      if (flags & (update_values | update_gradients))
+        out |= update_quadrature_points;
+
+      // if need gradients, add update_values due to product rule
+      if (out & update_gradients)
+        out |= update_values;
+    }
+
+  Assert (!(flags & update_3rd_derivatives),
+          ExcNotImplemented());
+
+  return out;
+}
+
+
+template <int dim, int spacedim>
+template <int dim_1>
+typename FiniteElement<dim,spacedim>::InternalDataBase *
+FE_Enriched<dim,spacedim>::setup_data (std::unique_ptr<typename FiniteElement<dim,spacedim>::InternalDataBase> fes_data,
+                                       const UpdateFlags      flags,
+                                       const Quadrature<dim_1> &quadrature) const
+{
+  Assert ((dynamic_cast<typename FESystem<dim,spacedim>::InternalData *> (fes_data.get()) != NULL),
+          ExcInternalError());
+  typename FESystem<dim,spacedim>::InternalData *data_fesystem =
+    static_cast<typename FESystem<dim,spacedim>::InternalData *> (fes_data.get());
+
+  // FESystem::InternalData will be aggregated (owned) by
+  // our InternalData.
+  fes_data.release();
+  InternalData *data = new InternalData(std::unique_ptr<typename FESystem<dim,spacedim>::InternalData>(data_fesystem));
+
+  // copy update_each from FESystem data:
+  data->update_each = data_fesystem->update_each;
+
+  // resize cache array according to requested flags
+  data->enrichment.resize(this->n_base_elements());
+
+  const unsigned int n_q_points = quadrature.size();
+
+  for (unsigned int base=0; base < this->n_base_elements(); ++base)
+    {
+      data->enrichment[base].resize(this->element_multiplicity(base));
+      for (unsigned int m = 0; m < this->element_multiplicity(base); ++m)
+        {
+          if (flags & update_values)
+            data->enrichment[base][m].values.resize(n_q_points);
+
+          if (flags & update_gradients)
+            data->enrichment[base][m].gradients.resize(n_q_points);
+
+          if (flags & update_hessians)
+            data->enrichment[base][m].hessians.resize(n_q_points);
+        }
+    }
+
+  return data;
+}
+
+
+template <int dim, int spacedim>
+typename FiniteElement<dim,spacedim>::InternalDataBase *
+FE_Enriched<dim,spacedim>::get_face_data (const UpdateFlags      update_flags,
+                                          const Mapping<dim,spacedim>    &mapping,
+                                          const Quadrature<dim-1> &quadrature,
+                                          internal::FEValues::FiniteElementRelatedData< dim, spacedim >        &output_data) const
+{
+  return setup_data(std::unique_ptr<typename FiniteElement<dim,spacedim>::InternalDataBase>(fe_system.get_face_data(update_flags,mapping,quadrature,output_data)),
```

Comment:
I don't think you need the explicit cast to `std::unique_ptr` here. The compiler should be able to do that for you.